port/utils/hex.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ * Copyright (c) 2024 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * These utilities already included in Zephyr tree but needed for other platforms for OTA service.
+ */
+
+#include <golioth/golioth_status.h>
+#include "hex.h"
+
+int char2hex(char c, uint8_t *x)
+{
+    if ((c >= '0') && (c <= '9'))
+    {
+        *x = c - '0';
+    }
+    else if ((c >= 'a') && (c <= 'f'))
+    {
+        *x = c - 'a' + 10;
+    }
+    else if ((c >= 'A') && (c <= 'F'))
+    {
+        *x = c - 'A' + 10;
+    }
+    else
+    {
+        return GOLIOTH_ERR_INVALID_FORMAT;
+    }
+
+    return 0;
+}
+
+size_t hex2bin(const char *hex, size_t hexlen, uint8_t *buf, size_t buflen)
```

Comment:
Thinking about this some more, I think it would make sense to convert the hash to binary as we download it, and store it as binary in the manifest struct. A user of the SDK would have no reason to expect a hash in ASCII format, it's just an artifact of the cloud API, and we shouldn't let that implementation detail dictate the Firmware API.

Reply:
Great idea, I've added a commit to store the hash as a binary array so it's ready for comparison.